### PR TITLE
Fix log rates

### PR DESCRIPTION
**Pull request by MMeent**
## Problem

We log a lot of things which don't always make sense to log.  This PR attempts to reduce the log traffic through various ways. See the commits for details.

## Summary of changes

- Stop logging HealthCheck message passing at INFO level (moved to DEBUG)
- Stop logging /status accesses at INFO (moved to DEBUG)
- Stop logging most `missing config file "compute_ctl_temp_override.conf"`
- Log memory usage only when the data has changed significantly, or if we've not recently logged the data, rather than always every 2 seconds.

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

**Review by mattpodraza — APPROVED** [review]

File: libs/vm_monitor/src/cgroup.rs [hunk 1]
```diff
@@ -115,12 +117,27 @@ impl CgroupWatcher {
...
+            //
+            // We skip logging the data if data hasn't meaningfully changed in a while, unless
+            // we've already ignored too many (=history_log_len) previous iterations.
             if i == history_log_len - 1 {
-                info!(
-                    history = ?MemoryStatus::debug_slice(&history_log_buf),
-                    summary = ?summary,
-                    "Recent cgroup memory statistics history"
-                );
+                if skipped > history_log_len
```
Comment (sharnoff) [inline libs/vm_monitor/src/cgroup.rs:124]: to clarify, now `history_log_len` is both:

1. the number of items that get logged in a single report, and
2. the maximum number of reports that can be skipped

?

File: libs/vm_monitor/src/dispatcher.rs [hunk 2]
```diff
@@ -118,7 +118,15 @@ impl Dispatcher {
...
     pub async fn send(&mut self, message: OutboundMsg) -> anyhow::Result<()> {
-        info!(?message, "sending message");
+        match &message.inner {
+            crate::protocol::OutboundMsgKind::HealthCheck { .. } => {
+                debug!(?message, "sending message");
+            }
+            _ => {
+                info!(?message, "sending message");
+            }
+        }
```
Comment (sharnoff) [inline libs/vm_monitor/src/dispatcher.rs:128]: thoughts on simplifying (here and elsewhere) with `if let ...` and/or `if matches!(...)` ?

just thinking ways to make this less verbose.

File: libs/vm_monitor/src/cgroup.rs [hunk 3]
```diff
@@ -232,6 +249,23 @@ impl MemoryStatus {
 
         DS(slice)
     }
+
+    /// Check if the other memory status is a close or similar result.
+    /// Returns true if the larger value is not larger than the smaller value
+    /// by 1/8 of the smaller value, and within 128MiB.
+    fn status_is_close_similar(&self, other: &MemoryStatus) -> bool {
```
Comment (tristan957) [inline libs/vm_monitor/src/cgroup.rs:256]: Could rename to:

- `status_is_close_or_similar`
- `status_has_meaningfully_changed`

Food for thought.

File: libs/vm_monitor/src/cgroup.rs [hunk 4]
```diff
@@ -232,6 +249,23 @@ impl MemoryStatus {
 
         DS(slice)
     }
+
+    /// Check if the other memory status is a close or similar result.
+    /// Returns true if the larger value is not larger than the smaller value
+    /// by 1/8 of the smaller value, and within 128MiB.
```
Comment (tristan957) [inline libs/vm_monitor/src/cgroup.rs:255]: Can't say the comment makes complete sense to me. Could you maybe add some doctests to this function, or examples?

File: compute_tools/src/http/api.rs [hunk 5]
```diff
@@ -48,7 +48,7 @@ async fn routes(req: Request<Body>, compute: &Arc<ComputeNode>) -> Response<Body
     match (req.method(), req.uri().path()) {
         // Serialized compute state.
         (&Method::GET, "/status") => {
-            info!("serving /status GET request");
+            debug!("serving /status GET request");
```
Comment (tristan957) [inline compute_tools/src/http/api.rs:51]: Perhaps it would make more sense to move this out of the match arm and just `debug!` every URI path + method that we receive, but overall the change is good.

File: libs/vm_monitor/src/cgroup.rs [hunk 6]
```diff
@@ -253,7 +254,8 @@ impl MemoryStatus {
     /// Check if the other memory status is a close or similar result.
     /// Returns true if the larger value is not larger than the smaller value
     /// by 1/8 of the smaller value, and within 128MiB.
-    fn status_is_close_similar(&self, other: &MemoryStatus) -> bool {
+    /// See
```
Comment (tristan957) [inline libs/vm_monitor/src/cgroup.rs:257]: Looks like the rest of this sentence is missing.

File: libs/vm_monitor/src/cgroup.rs [hunk 7]
```diff
@@ -232,6 +250,24 @@ impl MemoryStatus {
 
         DS(slice)
     }
+
+    /// Check if the other memory status is a close or similar result.
+    /// Returns true if the larger value is not larger than the smaller value
+    /// by 1/8 of the smaller value, and within 128MiB.
+    /// See check_similarity_behaviour for examples of behaviour
```
Comment (tristan957) [inline libs/vm_monitor/src/cgroup.rs:257]: You can make this a link in rustdoc: https://doc.rust-lang.org/rustdoc/write-documentation/linking-to-items-by-name.html

**Review by tristan957 — APPROVED** [review]
1 last comment.

**Review by tristan957 — APPROVED** [review]

File: libs/vm_monitor/src/cgroup.rs [hunk 8]
```diff
@@ -85,7 +88,10 @@ impl CgroupWatcher {
 
         // buffer for samples that will be logged. once full, it remains so.
         let history_log_len = self.config.memory_history_log_interval;
+        let max_skip = self.config.memory_history_log_noskip_interval;
         let mut history_log_buf = vec![MemoryStatus::zeroed(); history_log_len];
+        let mut last_logged_memusage = MemoryStatus::zeroed();
+        let mut can_skip_logs_until = Instant::now() - max_skip;
```
Comment (sharnoff) [inline libs/vm_monitor/src/cgroup.rs:94]: Why `now() - max_skip` ? (maybe I'm missing something, but won't `now()` suffice, because inside the loop `now()` will be newer than it?)

File: libs/vm_monitor/src/cgroup.rs [hunk 9]
```diff
@@ -115,12 +121,24 @@ impl CgroupWatcher {
             // equal to the logging interval, we can just log the entire buffer every time we set
             // the last entry, which also means that for this log line, we can ignore that it's a
             // ring buffer (because all the entries are in order of increasing time).
-            if i == history_log_len - 1 {
+            //
+            // We skip logging the data if data hasn't meaningfully changed in a while, unless
+            // we've already ignored too many (=history_log_len) previous iterations.
```
Comment (sharnoff) [inline libs/vm_monitor/src/cgroup.rs:126]: > (=history_log_len)

this needs updating now, right?

File: libs/vm_monitor/src/cgroup.rs [hunk 10]
```diff
@@ -33,6 +35,7 @@ impl Default for Config {
             memory_poll_interval: Duration::from_millis(100),
             memory_history_len: 5, // use 500ms of history for decision-making
             memory_history_log_interval: 20, // but only log every ~2s (otherwise it's spammy)
+            memory_history_log_noskip_interval: Duration::from_secs(60), // but only if it's changed, or 60 seconds have passed
```
Comment (sharnoff) [inline libs/vm_monitor/src/cgroup.rs:38]: let's put this more frequent, at e.g. 15s ? That way it's not lower resolution than our metrics (ext-metrics datasource is 15s IIRC). If it's still a substantial log volume, I'd be happy to reduce it further.

File: libs/vm_monitor/src/cgroup.rs [hunk 11]
```diff
@@ -53,7 +56,7 @@ pub struct CgroupWatcher {
 
 impl CgroupWatcher {
     /// Create a new `CgroupWatcher`.
-    #[tracing::instrument(skip_all, fields(%name))]
+    #[tracing::instrument(skip_all, fields(% name))]
```
Comment (sharnoff) [inline libs/vm_monitor/src/cgroup.rs:59]:
```suggestion
    #[tracing::instrument(skip_all, fields(%name))]
```

to avoid unnecessary diff and remain consistent with the codebase

File: libs/vm_monitor/src/dispatcher.rs [hunk 12]
```diff
@@ -118,7 +119,12 @@ impl Dispatcher {
     /// serialize the wrong thing and send it, since `self.sink.send` will take
     /// any string.
     pub async fn send(&mut self, message: OutboundMsg) -> anyhow::Result<()> {
-        info!(?message, "sending message");
+        if matches!(&message.inner, HealthCheck { .. }) {
```
Comment (sharnoff) [inline libs/vm_monitor/src/dispatcher.rs:122]: For consistency, I would recommend just importing `OutboundMsgKind` and then use `OutboundMsgKind::HealthCheck` here (that way, it gets added to the `use crate::protocol::{ ... }`)

**Review by sharnoff — APPROVED** [review]
a handful of small comments asking for minor changes; otherwise, looks good!

File: libs/vm_monitor/src/cgroup.rs (hunk 8 above)
Comment (sharnoff) [reply libs/vm_monitor/src/cgroup.rs:94]: Ah, I see - because `Instant`s are only guaranteed to be non-decreasing?

Let's add a brief comment saying that?

File: libs/vm_monitor/src/cgroup.rs (hunk 10 above)
Comment (sharnoff) [reply libs/vm_monitor/src/cgroup.rs:38]: My main concern is less that the data isn't available elsewhere, it's more about not needing to cross-compare different datasources. Add onto it that metrics are imprecise (if there's datapoints every 15s, I don't know when the data was actually collected in that 15-second interval), and I've typically found logs to be much more helpful for things like investigating OOMs.

Comment (sharnoff) [reply libs/vm_monitor/src/cgroup.rs]: > Isn't that implied by the timestamp associated with the collection of the metric?

My understanding is that this is unfortunately not quite the case — for example, if there's data every 15s, you will get datapoints at 00:00:00, 00:00:15, 00:00:30, etc. But they might've been collected at 23:59:53, 00:00:08, 00:00:23, etc.

(again: this is just "as far as I can tell" -- I'm by no means an expert here. A while back I tried to make a compute metrics dashboard that used cAdvisor's `container_last_seen` metric, which _does_ have the actual timestamp used, and it was a bit off from the time associated with the value)
